pkg/controller/traceflow/validate.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traceflow
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+
+	admv1 "k8s.io/api/admission/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/features"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+func (c *Controller) Validate(review *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	newResponse := func(allowed bool, deniedReason string) *admv1.AdmissionResponse {
+		resp := &admv1.AdmissionResponse{
+			UID:     review.Request.UID,
+			Allowed: allowed,
+		}
+		if !allowed {
+			resp.Result = &metav1.Status{
+				Message: deniedReason,
+			}
+		}
+		return resp
+	}
+
+	klog.V(2).InfoS("Validating Traceflow", "request", review.Request)
+
+	var newObj, oldObj crdv1alpha1.Traceflow
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current Traceflow")
+			return newResponse(false, err.Error())
+		}
+	}
+	if review.Request.OldObject.Raw != nil {
+		if err := json.Unmarshal(review.Request.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing old Traceflow")
+			return newResponse(false, err.Error())
+		}
+	}
+
+	switch review.Request.Operation {
+	case admv1.Create:
+		klog.V(2).InfoS("Validating CREATE request for Traceflow")
+		allowed, deniedReason := c.validateCreate(&newObj)
+		return newResponse(allowed, deniedReason)
+	case admv1.Update:
+		klog.V(2).InfoS("Validating UPDATE request for Traceflow")
+		allowed, deniedReason := c.validateUpdate(&oldObj, &newObj)
+		return newResponse(allowed, deniedReason)
+	case admv1.Delete:
+		// This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.
+		klog.InfoS("Validating DELETE request for Traceflow. This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.")
+		return newResponse(true, "")
+	default:
+		klog.InfoS("Unknown request type for Traceflow")
+		return newResponse(false, "unknown request type for Traceflow")
+	}
+}
+
+func (c *Controller) validateCreate(tf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	if tf.Spec.Destination.Service != "" && !features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
```

Comment:
We don't require AntreaProxy to be enabled for antrea-controller, so this won't reflect whether the feature is really enabled in antrea-agent.
https://github.com/antrea-io/antrea/blob/main/docs/feature-gates.md#list-of-available-features

Since there is still a validation in agent, this can be removed.

Reply:
I've moved this validation back to the Agent.